README.md
Bookmarker UI is a React component library that implements Neumorphic Design, which includes a collection of components ready for production.

# Contents

- [Contents](#contents)
- [Introduction](#introduction)
- [Getting started](#getting-started)
  - [Installation](#installation)
  - [Provider](#provider)
  - [Theme](#theme)
- [Components](#components)
  - [Avatar](#avatar)
  - [Board](#board)
  - [Button](#button)
  - [Checkbox](#checkbox)
  - [Chip](#chip)
  - [Container](#container)
  - [Figure](#figure)
  - [Header](#header)
  - [Hide](#hide)
  - [Input](#input)
  - [Link](#link)
  - [Loader](#loader)
  - [Radio](#radio)
  - [Select](#select)
  - [Textarea](#textarea)
- [Props](#props)
- [Hooks](#hooks)
  - [useWindowSize](#usewindowsize)
  - [useClickAway](#useclickaway)
- [Scripts](#scripts)
  - [darken](#darken)
  - [contrastColor](#contrastcolor)

# Introduction

This is a basic react components using neumorphic styles. For visualisation, please check out storybook [here](https://bookmarker-ui.vercel.app)

# Getting started

## Installation

```sh
  npm i bookmarker-ui
```

It is built using styled-components as CSS-in-JS, so it is required for it to work.

```sh
  npm i styled-components
```

## Provider

Use ThemeProvider at the root of your app, or on the ancestor of any components that you may need it.

```tsx
import { ThemeProvider } from 'bookmarker-ui';

export default function index() {
  return (
    <ThemeProvider>
      <App />
    </ThemeProvider>
  );
}
```

## Theme

`bookmarker-ui` comes with its own theme, but you may want to update it. This is possible inside `ThemeProvider`.

```tsx
import { ThemeProvider, theme, ThemeProps } from 'bookmarker-ui';

const myTheme: ThemeProps = {
  ...theme,
  palette: {
    ...theme.palette,
    primary: {
      main: '#ade01a',
      dark: '#8cb515',
      light: '#bff71e',
    },
  },
};
export default function index() {
  return (
    <ThemeProvider theme={theme}>
      <App />
    </ThemeProvider>
  );
}
```

See [ThemeProps](#themeprops) for full details about the props

# Components

## Avatar

```tsx
import { Avatar } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Avatar src="https://picsum.photos/200/300" />
      <Avatar
        shape="rounded"
        src="https://picsum.photos/200/300"
        style={{
          height: 200,
        }}
      />
      <Avatar
        shape="circle"
        src="https://picsum.photos/200/300"
        style={{
          height: 200,
        }}
      />
    </div>
  );
}
```

## Board

```tsx
import { Board } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Board depth={1}>
        <div style={{ padding: 8 }}>Low</div>
      </Board>
      <Board depth={2}>
        <div style={{ padding: 8 }}>Medium</div>
      </Board>
      <Board depth={3}>
        <div style={{ padding: 8 }}>Deep</div>
      </Board>
    </div>
  );
}
```

## Button

This is a basic button, for interaction or clicking

```tsx
import { Button } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Button>Default</Button>
      <Button variant="primary">Primary</Button>
      <Button variant="secondary">Secondary</Button>
      <Button disabled>Default</Button>
      <Button disabled variant="primary">
        Primary
      </Button>
      <Button disabled variant="secondary">
        Secondary
      </Button>
    </div>
  );
}
```

## Checkbox

This is a basic button, for interaction or clicking

```tsx
export default function Demo() {
  return (
    <div>
      <Checkbox label="Default" />
      <Checkbox checked label="Primary" variant="primary" />
      <Checkbox checked label="Secondary" variant="secondary" />
      <Checkbox disabled label="Disabled" />
    </div>
  );
}
```

## Chip

Used to indicate labels or tags

```tsx
import { Chip } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Chip title="Default" />
      <Chip disabled title="Default disabled" />
      <Chip onClick={() => console.log('hello')} title="Clickable" />
      <Chip
        disabled
        onClick={() => console.log('hello')}
        title="Clickable and Disabled"
      />
    </div>
  );
}
```

## Container

```tsx
import { Container } from 'bookmarker-ui';

export default function Demo() {
  return <Container>This is the body</Container>;
}
```

## Figure

```tsx
import { Avatar, Figure, FigCaption } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Figure caption="Hello Image" src="https://picsum.photos/200/300" />
      <Figure
        caption="Hello Image"
        src="https://picsum.photos/200/300"
        shape="rounded"
      />
      <Figure
        caption="Hello Image"
        src="https://picsum.photos/200/300"
        shape="circle"
      />
    </div>
  );
}
```

## Header

```tsx
import { MenuItem, Hide, Input, Header } from 'bookmarker-ui';

export default function Demo() {
  return (
    <Header style={{ justifyContent: 'space-between' }}>
      <MenuItem active>
        <img
          src="https://bookmarker-one.vercel.app/favicon.ico"
          style={{ height: 24, width: 24 }}
        />
      </MenuItem>
      <Hide smDown>
        <Input
          onChange={() => {}}
          placeholder="filter"
          type="search"
          value=""
        />
      </Hide>
      <div style={{ display: 'flex', flexDirection: 'row' }}>
        <MenuItem active>Profile</MenuItem>
        <MenuItem>Sign out</MenuItem>
      </div>
    </Header>
  );
}
```

## Hide

```tsx
import { Hide } from 'bookmarker-ui';

export default function Demo() {
  const { width } = useWindowSize();

  return (
    <div
      style={{
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      <div>size: {width}</div>
      <Hide xsDown>xsDown: 320</Hide>
      <Hide xsUp>xsUp: 320</Hide>
      <Hide smDown>smDown: 640</Hide>
      <Hide smUp>smUp: 640</Hide>
      <Hide mdDown>mdDown: 960</Hide>
      <Hide mdUp>mdUp: 960</Hide>
      <Hide lgDown>lgDown: 1280</Hide>
      <Hide lgUp>lgUp: 1280</Hide>
      <Hide xlDown>xlDown: 1920</Hide>
      <Hide xlUp>xlUp: 1920</Hide>
    </div>
  );
}
```

## Input

```tsx
import { Input } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Input value="Default" />
      <Input disabled value="Default" />
      <Input placeholder="Write stuff here" value="" />
      <Input
        errorText="Error text here"
        placeholder="Write stuff here"
        value=""
      />
      <Input label="Email" placeholder="example@email.com" value="" />
    </div>
  );
}
```

## Link

```tsx
import { Link } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Link>Default</Link>
      <Link href="https://adeolaade.com/">Repo</Link>
    </div>
  );
}
```

## Loader

```tsx
import { Loader } from 'bookmarker-ui';

export default function Demo() {
  return <Loader />;
}
```

## Radio

This is a basic button, for interaction or clicking

```tsx
export default function Demo() {
  return (
    <div>
      <Radio
        wrapperStyle={{
          marginBottom: 8,
        }}
        label="Default"
      />
      <Radio checked label="Primary" variant="primary" />
      <Radio checked label="Secondary" variant="secondary" />
      <Radio disabled label="Disabled" />
    </div>
  );
}
```

## Select

```tsx
export default function Demo() {
  const [value, setValue] = useState('');
  const list = ['One', 'Two', 'Three', 'Four', 'Five'];

  return (
    <Select title={value || 'Select a number'}>
      {list.map(item => (
        <SelectItem
          onClick={() => setValue(item)}
          active={value === item}
          key={item}
        >
          {item}
        </SelectItem>
      ))}
    </Select>
  );
}
```

## Textarea

```tsx
import { Textarea } from 'bookmarker-ui';

export default function Demo() {
  return (
    <div>
      <Textarea rows={4} value="Default" />
      <Textarea disabled rows={4} value="Default" />
      <Textarea placeholder="Write stuff here" rows={4} value="" />
      <Textarea
        errorText="Error just occurred"
        placeholder="Write stuff here"
        rows={4}
        value=""
      />
      <Textarea
        label="Email"
        placeholder="example@email.com"
        rows={4}
        value=""
      />
    </div>
  );
}
```

# Props

- ThemeProps
- BoardProps
- ButtonProps
- ChipProps
- HeaderProps
- HideProps
- InputProps
- LinkProps
- TextareaProps

# Hooks

## useWindowSize

```ts
import { useWindowSize } from 'bookmarker-ui';

const Demo = () => {
  const { height, width } = useWindowSize();
  console.log(height, '==', width);
  // 1000 == 400

  return <div>This is it</div>;
};
```

## useClickAway

```tsx
import { useClickAway } from 'bookmarker-ui';

const Demo = () => {
  const ref = useRef(null);

  useClickAway(ref, () => {
    console.log('Outside clicked');
  });

  return <div ref={ref}>Hello there</div>;
};
```

# Scripts

## darken

> same as lighten

```ts
// default is 0.2
darken('#ade01a'); // #8AB315
darken('#ade01a', 0.1); // #9CCA17
darken('#ade01a', 0.3); // #799D12
darken('#ade01a', 0.5); // #57700D
darken('#ade01a', 0.7); // #344308
darken('#ade01a', 0.9); // #111603
darken('#ade01a', 1); // #000000
```

## contrastColor

Returns white or black, depending on the colour provided. It will determine the best colour for contrast.

```ts
contrastColor('#fff'); //"#000000"
contrastColor('#000'); //"#ffffff"
contrastColor('#ade01a'); //"#000000"
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

- MAJOR.MINOR.PATCH

1. Major: Incompatible changes or change how the system works
2. Minor: Add new components or functionalities
3. Patch: Bug fixes or adjustment to existing component

## [Unreleased]

Below are some of the things that I'm working on or plan to add.

### Add

- Switch
- Radio button
- Tabs
- Select
- Rating
- Avatar
- Divider
- Icons
- Table
- Divider
- Transition: Fade, Slide (up, down, left and right), Zoom
- Modal
- Rating
- Select
- Slider
- Icons
- Icon button
- List
- Table
- Notify
- Skeleton
- Breadcrumbs
- Drawer

## [1.1.2@alpha] - 2022-10-17

### Added

- Footer component
- Up arrow added

### Changed

- Resize headings
- Add smooth scrolling
- Update color scheme (shades)
- Caret icon updated

### Fixed

- Add props to text variant of Typography

## [1.1.1@alpha] - 2022-10-17

### Added

- Fix stroke width

## [1.1.0] - 2022-10-17

### Added

- Figure created
- Avatar created

### Changed

- close select upon click

## [1.0.2] - 2022-10-17

### Added

- Checkbox
- Radio

### Changed

- icon extracted from react-icons
- Change elevated box-shadow to white

## [1.0.1] - 2022-10-17

### Changed

- Paths updated
- Make use of svg in loader

## Fixed

- U-turn on the paths

## [0.8.0] - 2022-10-11

### Add

- Typography
- Select added

### Changed

- Card changed to Board
- Loader changed

## [0.7.0] - 2022-10-10

### Add

- Card added

### Changed

- disabled in chips set to optional

## [0.6.2] - 2022-10-10

### Changed

- make loader's height dynamic with `fullHeight` props

## [0.6.1] - 2022-10-10

### Changed

- make container's height dynamic with `fullHeight` props
- add centering props to container with `alignCenter` and `justifyCenter` props

## [0.6.0] - 2022-10-08

### Added

- Pre-commit checks added
- Changelog added

### Changed

- changed loader image from `.gif` to `data-url`

### Fixed

- Positioning in storybook

## [0.5.0] - 2022-10-05

### Added

- Container added
- Loader added

## [0.4.0] - 2022-10-04

### Added

- UI page created [here](https://bookmarker-one.vercel.app/)
- Input added
- Links added
- Textarea added

## [0.2.0] - 2022-10-03

### Added

- README added
- Chips added

### Changed

- Theme interface changed to ThemeProps
- Stories updated

## [0.1.2] - 2022-10-01

### Changed

- Button added to root for import

### Added

- Styled-components added
- Button created
- Theme added


src/theme/index.ts
import ThemeProps from './ThemeProps';
import { darken, lighten } from '../scripts/color-functions';
import ThemeProvider from './ThemeProvider';

const breakpoint = {
  xs: 320,
  sm: 640,
  md: 960,
  lg: 1280,
  xl: 1920,
};

const primary = {
  main: '#00796b',
  dark: '#004d40',
  light: '#4db6ac',
};

const secondary = {
  main: '#e91e63',
  dark: '#c2185b',
  light: '#f06292',
};

const error = {
  main: '#e83c3d',
  dark: darken('#e83c3d'),
  light: lighten('#e83c3d'),
};

const warning = {
  main: '#ffac03',
  dark: darken('#ffac03'),
  light: lighten('#ffac03'),
};

const success = {
  main: '#329b52',
  dark: darken('#329b52'),
  light: lighten('#329b52'),
};

const stack = {
  appbar: 1100,
  drawer: 1200,
  modal: 1300,
  snackbar: 1400,
  tooltip: 1500,
};

const duration = {
  short: '0.4s ease 0s',
  medium: '0.5s ease 0s',
  long: '0.6s ease 0s',
};

const main = {
  black: '#000000',
  white: '#ffffff',
};

const shade = {
  lightest: '#e4ebf5', // greyLight1
  lighter: '#c8d0e7', // greyLight2
  light: '#bec8e4', // greyLight3
  dark: '#9baacf', // greyDark
  darker: '#8694b5', //
  darkest: '#485575', //
};

const elevation = {
  low: `${shade.light} 0.15rem 0.15rem 0.3rem, ${main.white} -0.1rem -0.1rem 0.25rem;`,
  medium: `${shade.light} 0.3rem 0.3rem 0.6rem, ${main.white} -0.2rem -0.2rem 0.5rem`,
  high: `${shade.light} 0.6rem 0.6rem 1.2rem, ${main.white} -0.4rem -0.4rem 1rem`,
};

const depth = {
  shallow: `${shade.lighter} 0.1rem 0.1rem 0.125rem inset, ${main.white} -0.1rem -0.1rem 0.125rem inset`,
  medium: `${shade.lighter} 0.2rem 0.2rem 0.25rem inset, ${main.white} -0.2rem -0.2rem 0.25rem inset`,
  deep: `${shade.lighter} 0.3rem 0.3rem 0.375rem inset, ${main.white} -0.3rem -0.3rem 0.375rem inset`,
};

export const theme: ThemeProps = {
  breakpoint,
  elevation,
  depth,
  stack,
  duration,
  shade,
  palette: {
    primary,
    secondary,
    error,
    warning,
    success,
    main,
  },
};

export { ThemeProps, ThemeProvider };
export default theme;
